platform/mbed_retarget.cpp

```diff
@@ -205,6 +213,141 @@ short DirectSerial::poll(short events) const
 }
 #endif
 
+#ifndef CIRCULAR_BUFFER_FILE_DEPTH
+#define CIRCULAR_BUFFER_FILE_DEPTH 512
+#endif
+
+class CircularBufferFile : public FileHandle {
+public:
+    virtual ~CircularBufferFile();
+    virtual ssize_t write(const void *buffer, size_t size);
+    virtual ssize_t read(void *buffer, size_t size);
+    virtual short poll(short events) const;
+    virtual off_t seek(off_t offset, int whence = SEEK_SET)
+    {
+        return -ESPIPE;
+
+    }
+    virtual off_t size()
+    {
+        return _buffer.size();
+    }
+    virtual int isatty()
+    {
+        return false;
+    }
+    virtual int close()
+    {
+        return 0;
+    }
+    virtual void sigio(Callback<void()> func);
+private:
+    /** Acquire mutex */
+    virtual void api_lock(void);
+
+    /** Release mutex */
+    virtual void api_unlock(void);
+    
+    void wait_ms(uint32_t millisec);
+private:
+    CircularBuffer<char, CIRCULAR_BUFFER_FILE_DEPTH> _buffer;
+    Callback<void()> _sigio_cb;
+    PlatformMutex _mutex;
+};
+
+CircularBufferFile::~CircularBufferFile(){}
+void CircularBufferFile::api_lock(void)
+{
+    _mutex.lock();
+}
+
+void CircularBufferFile::api_unlock(void)
+{
+    _mutex.unlock();
+}
+
+ssize_t CircularBufferFile::write(const void* buffer, size_t size) {
+    const char* b = static_cast<const char*>(buffer);
+    if (size == 0) {
+        return 0;
+    }
+    api_lock();
+    for ( size_t i = 0; i < size; i++){
+        _buffer.push(b[i]);
+    }
+    api_unlock();
+    size_t data_written =  size  % CIRCULAR_BUFFER_FILE_DEPTH;
```

Comment:
`data_written` is always `size`. You didn't partially consume the data - you consumed it all and maybe threw some away.

If you're implementing the sigio callback, this should be calling it if writing makes the device readable. If you're not going to implement non-blocking function, you could just avoid implementing sigio, I guess.